### DISCUSSION TITLE
Error Handling

### DISCUSSION THREAD

**Discussion by MaheshkumarSundaram** [bpmnServer/bpmn-server#220] [Q&A] [answered]
Hi @ralphhanna

Reagrding error handling, consider the below example:

![image](https://github.com/user-attachments/assets/25b7e048-206a-40f3-9edd-d424aa1666dd)

For testing, I have the below service:

```js
    async ErrorHandleTest(input, context) {
        if (Math.random() < 0.1) {
            return { bpmnError: 'Something wrong' }; // my required ERROR message
        } else {
           // no error
            return;
        }
    }
```

Where is that error and its message stored and how to access it?

**Comment by ralphhanna** [comment]
Not anywhere, but if the error handler is error specific it can store it anywhere.

Any suggestions?

One possible change:
code before:
```ts
    async processError(errorCode,callingEvent) {

        let errorHandlerToken = this.getScopeCatchEvent('error',errorCode);
        if (errorHandlerToken) {
            await errorHandlerToken.signal(null);
            this.currentItem.status = ITEM_STATUS.end;
            await this.end(true);
        }
        else {
            this.log("Aborting due to error");
            await this.execution.terminate();
            return;
        }
    }

```
Code After:
```ts
    async processError(errorCode,callingEvent) {

        let errorHandlerToken = this.getScopeCatchEvent('error',errorCode);
        if (errorHandlerToken) {
            await errorHandlerToken.signal(errorCode);
            this.currentItem.status = ITEM_STATUS.end;
            await this.end(true);
        }
        else {
            this.log("Aborting due to bpmnError",errorCode)
            await appDelegate.bpmnError(errorCode,execution); //this.execution.terminate();
            return;
        }

    }
    
```

**Comment by MaheshkumarSundaram** [comment]
@ralphhanna 

Wouldn't it better if the error message/code is automatically stored in the catch event (probably `vars` but not sure whether `vars` would be ideal for this though!!!)? In case of not being caught, automatically storing the error message/code in the service or script task itself. 

This would be helpful to see what went wrong when coming back to that item.

what do you think?

**Reply by MaheshkumarSundaram** [reply]
@ralphhanna, any thoughts on this?

**Comment by ralphhanna** [comment, marked as answer]
@MaheshkumarSundaram 
Hi, After long thoughts, I have added a new variable `item.statusDetails` that contains and object with bpmnError information.

Please check new release 2.2.10 

Thanks

**Reply by MaheshkumarSundaram** [reply]
Thanks @ralphhanna.
